c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,378 @@
+        _ => None,
+    }
+}
+
+fn get_assign_sides<'tcx, 'a>(
+    stmt: &'a Statement<'tcx>,
+) -> Option<(Place<'tcx>, &'a Rvalue<'tcx>)> {
+    let (pl, ref rv) = match &stmt.kind {
+        StatementKind::Assign(it) => Some(&**it),
+        _ => None,
+    }?;
+    Some((*pl, rv))
+}
+
+impl<'tcx> CVoidPtr<'tcx> {
+    /// Try to get the appropriate [`CVoidCast`]
+    /// from this [`Statement`] in the given [`CVoidCastDirection`].
+    ///
+    /// For the [`From`] direction, the [`*c_void`] is on the rhs.\
+    /// For the [`To`] direction, the [`*c_void`] is on the lhs.
+    ///
+    /// [`*c_void`]: core::ffi::c_void
+    /// [`From`]: CVoidCastDirection::From
+    /// [`To`]: CVoidCastDirection::To
+    pub fn get_cast_from_stmt(
+        &self,
+        direction: CVoidCastDirection,
+        stmt: &Statement<'tcx>,
+    ) -> Option<CVoidCast<'tcx>> {
+        let (lhs, rv) = get_assign_sides(stmt)?;
+        let rhs = get_cast_place(rv)?;
+
+        use CVoidCastDirection::*;
+        let cast = match direction {
+            From => CVoidCast {
+                c_void_ptr: self.checked_by_eq(rhs).ok()?,
+                other_ptr: lhs,
+            },
+            To => CVoidCast {
+                c_void_ptr: self.checked_by_eq(lhs).ok()?,
+                other_ptr: rhs,
+            },
+        };
+        Some(cast)
+    }
+}
+
+impl<'tcx> Borrow<Place<'tcx>> for CVoidPtr<'tcx> {
+    fn borrow(&self) -> &Place<'tcx> {
+        &self.place
+    }
+}
+
+/// A cast to/from a [`*c_void`](core::ffi::c_void) to/from a properly typed pointer.
+pub struct CVoidCast<'tcx> {
+    /// The [`*c_void`](core::ffi::c_void) side of the cast.
+    ///
+    /// It can be either the [`From`] or [`To`] side.
+    ///
+    /// [`From`]: CVoidCastDirection::From
+    /// [`To`]: CVoidCastDirection::To
+    c_void_ptr: CVoidPtr<'tcx>,
+
+    /// The non-[`*c_void`](core::ffi::c_void), properly typed pointer side of the cast.
+    ///
+    /// It can be either the [`From`] or [`To`] side.
+    ///
+    /// [`From`]: CVoidCastDirection::From
+    /// [`To`]: CVoidCastDirection::To
+    other_ptr: Place<'tcx>,
+}
+
+/// A mapping from [`*c_void`](core::ffi::c_void)s ([`CVoidPtr`]s)
+/// to their properly typed pointers,
+/// like [`CVoidCasts`], but in a single direction, meaning it represents either
+/// [`CVoidCastDirection::From`] or [`CVoidCastDirection::To`].
+#[derive(Default)]
+pub struct CVoidCastsUniDirectional<'tcx>(HashMap<CVoidPtr<'tcx>, Place<'tcx>>);
+
+impl<'tcx> CVoidCastsUniDirectional<'tcx> {
+    pub fn contains(&self, place: Place<'tcx>) -> bool {
+        self.0.contains_key(&place)
+    }
+
+    /// Get the adjusted [`Place`], skipping over [`*c_void`](core::ffi::c_void) intermediaries.
+    ///
+    /// That is, if `place` is a [`CVoidPtr`] in this map of [`CVoidCast`]s,
+    /// then the [`Place`] of its other, property-typed pointer is returned.
+    /// Otherwise, the same `place` is returned, as no adjustmenets are necessary.
```

Comment:
`adjustments`

Reply:
Fixed in 278df7d9361f7d9815104deca07351c5304e3a03 (`rebase` `fixup`).